doc/decisions/1_in_discussion/internal_cache.md

```diff
@@ -157,9 +197,98 @@ ksRemoveByName (cowMeta, "meta:/type");
 > I disagree, it is actually the same kind of surprise for "More Keys".
 > Only the "Fewer Keys" would get fixed.
 
-## Decision
+### Full-blown copy-on-write implementation
+
+Make Elektras `Key` and `KeySet` datastructures copy-on-write.
+This requires some major refactoring of code within `libelektra-core`.
+Code that does only interact with the datastructures via the public `libelektra-core` API should not notice any differences.
+The `mmapstorage` plugin will need a major refactoring.
+
+### Changes to `Key`
+
+For the `Key`, we need to extract everything for the data and name into their own structs.
+This is done for memory-management reasons, as we need to track how many keys point to the same data and/or name.
+
+In the data structures below, an empty key does have 96 bytes.
+An empty key of the current implementation has 64 bytes.
+
+A copied, non-modified key with the data structure below does always have 32 bytes.
+A copied, non-modified key of the current implementation has at least 64 bytes (for an empty key), but in reality much more as the name and the data are also copied.
+
+```c
+struct _KeyData {
+    union {
+        char * c;
+        void * v;
+    } data;
+
+    size_t dataSize;
+
+    uint16_t refs;
+    uint16_t reserved;
+};
+
+struct _KeyName {
+    char * key;
+    size_t keySize;
+
+    char * ukey;
+    size_t keyUSize;
+
+    uint16_t refs;
+    uint16_t reserved;
+};
+
+struct _Key {
+    struct _KeyData * keyData;    
+    struct _KeyName * keyName;
+    KeySet * meta;
+    keyflag_t flags;
+
+    uint16_t refs;
+    uint16_t reserved;
+};
+```
+
+### Changes to `KeySet`
 
-Not yet decided.
+For `KeySet`, we need to split out everything to do with the stored keys into a separate datastructure.
```

Comment:
Two things,

1. If we delete a copied KeySet, we don't known which KeySet is the original, so we couldn't decrement the counter. This could be dealt with storing a pointer to the original KeySet.

2. If the original KeySet is deleted, we don't know which other KeySets point at the data, so updating their count would not work

3. In similar fashion, if you update the original KeySet, the copied KeySets will also contain the new data (if the memory address does not change). This is unexpected behaviour.

Reply:
[9c035ac](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/9c035ac17415843d155e1cca7159b1ff39dedc93)